classes/table/accessaudit_table.php

```diff
@@ -0,0 +1,258 @@
+            'fullname' => get_string('fullname'),
+            'role' => get_string('role'),
+            'context' => get_string('report_context', 'report_accessaudit'),
+            'contextsystem' => get_string('report_contextsystem', 'report_accessaudit'),
+            'contextuser' => get_string('report_contextuser', 'report_accessaudit'),
+            'contextcoursecat' => get_string('report_contextcoursecat', 'report_accessaudit'),
+            'contextcourse' => get_string('report_contextcourse', 'report_accessaudit'),
+            'contextmodule' => get_string('report_contextmodule', 'report_accessaudit'),
+            'contextblock' => get_string('report_contextblock', 'report_accessaudit')
+        );
+
+        $this->define_columns(array_keys($cols));
+        $this->define_headers(array_values($cols));
+    }
+
+    /**
+     * Define table configs.
+     *
+     * @param \moodle_url $url url of the page where this table would be displayed.
+     */
+    protected function define_table_configs(\moodle_url $url) {
+        $urlparams = (array)$this->filters;
+
+        unset($urlparams['submitbutton']);
+
+        $url->params($urlparams);
+        $this->define_baseurl($url);
+
+        // Set table configs.
+        $this->collapsible(false);
+        $this->sortable(false);
+        $this->pageable(false);
+
+        $this->is_downloadable(true);
+        $this->show_download_buttons_at([TABLE_P_BOTTOM]);
+    }
+
+    /**
+     * Query the reader. Store results in the object for use by build_table.
+     *
+     * @param int $pagesize size of page for paginated displayed table.
+     * @param bool $useinitialsbar do you want to use the initials bar.
+     */
+    public function query_db($pagesize, $useinitialsbar = true) {
+        global $DB;
+
+        $offset = $pagesize * $this->page;
+        $limit = $pagesize;
+
+        list($countsql, $countparams) = $this->get_sql_and_params(true);
+        list($sql, $params) = $this->get_sql_and_params();
+
+        $total = $DB->count_records_sql($countsql, $countparams);
+
+        if ($this->is_downloading()) {
+            $users = $DB->get_records_sql($sql, $params);
+        } else {
+            $users = $DB->get_records_sql($sql, $params, $offset, $limit);
+        }
+
+        foreach ($users as $user) {
+            $data = new \stdClass();
+            $data->id = $user->id;
+            $data->idnumber = $user->idnumber;
+            $data->username = $user->username;
+            $data->email = $user->email;
+            $data->firstname = $user->firstname;
+            $data->lastname = $user->lastname;
+
+            if ($user->contextid > 0) {
+                $context = \context_helper::instance_by_id($user->contextid);
+                $contextids = explode('/', trim($context->path, '/'));
+                foreach ($contextids as $contextid) {
+                    $eachcontext = \context_helper::instance_by_id($contextid);
```

Comment:
This loop in a loop will probably be a perf bottle neck, but lets see it in action and then optimize it later

Reply:
Actually when you convert the results to a recordset you might need to handle this. see if it is possible to do the context breakdown in sql

Reply:
Hi @brendanheywood 
I have tried ```    $select = "u.id, u.idnumber, u.firstname, u.lastname, u.username,
            u.email, ra.contextid, r.shortname, ct.contextlevel, ct.path, ct.depth,
            split_part(ct.path, '/', 1) AS part1,
            split_part(ct.path, '/', 2) AS part2,
            split_part(ct.path, '/', 3) AS part3,
            split_part(ct.path, '/', 4) AS part4,
            split_part(ct.path, '/', 5) AS part5,
            split_part(ct.path, '/', 6) AS part6";```  since the depth is not a certain number. We still need a for loop to get the context instance and then put them in the right column. Do you have better idea? thanks